indra/llcommon/lua_function.cpp

```diff
@@ -974,6 +1002,289 @@ lua_function(
+        "    meta.__", name, " = function(lhs, rhs)\n"
+        "        if getmetatable(lhs) == meta then\n"
+        "            return lhs._target ", op, " rhs\n"
+        "        else\n"
+        "            return lhs ", op, " rhs._target\n"
+        "        end\n"
+        "    end\n");
+}
+
+// setdtor_refs __index() metamethod
+int setdtor_refs::meta__index(lua_State* L)
+{
+    // called with (setdtor_refs userdata, key), returns retrieved object
+    lua_checkdelta(L, -1);
+    lluau_checkstack(L, 2);
+    // stack: proxy, key
+    // get ptr to the C++ struct data
+    auto ptr = lua_toclass<setdtor_refs>(L, -2);
+    // meta__index() should NEVER be called with anything but setdtor_refs!
+    llassert(ptr);
+    // push the wrapped object
+    lua_getref(L, ptr->objref);
+    // stack: proxy, key, _target
+    // replace userdata with _target
+    lua_replace(L, -3);
+    // stack: _target, key
+    // Duplicate key because lua_tostring() converts number to string:
+    // if the key is (e.g.) 1, don't try to retrieve _target["1"]!
+    lua_pushvalue(L, -1);
+    // stack: _target, key, key
+    // recognize the special _target field
+    if (lua_tostdstring(L, -1) == "_target")
+    {
+        // okay, ditch both copies of "_target" string key
+        lua_pop(L, 2);
+        // stack: _target
+    }
+    else                            // any key but _target
+    {
+        // ditch stringized key
+        lua_pop(L, 1);
+        // stack: _target, key
+        // replace key with _target[key], invoking metamethod if any
+        lua_gettable(L, -2);
+        // stack: _target, _target[key]
+        // discard _target
+        lua_remove(L, -2);
+        // stack: _target[key]
+    }
+    return 1;
+}
+
+// When Lua destroys a setdtor_refs userdata object, either from garbage
+// collection or from LL.atexit(lua_destroybounduserdata), it's time to keep
+// its promise to call the specified Lua destructor function with the
+// specified Lua object. Of course we must also delete the captured
+// "references" to both objects.
+setdtor_refs::~setdtor_refs()
+{
+    lua_checkdelta(L);
+    lluau_checkstack(L, 2);
+    // push Registry[dtorref]
+    lua_getref(L, dtorref);
+    // push Registry[objref]
+    lua_getref(L, objref);
+    // free Registry[dtorref]
+    lua_unref(L, dtorref);
+    // free Registry[objref]
+    lua_unref(L, objref);
+    // call dtor(obj): one arg, no result, no error function
+    int rc = lua_pcall(L, 1, 0, 0);
+    if (rc != LUA_OK)
+    {
+        // TODO: we don't really want to propagate the error here.
+        // If this setdtor_refs instance is being destroyed by
+        // LL.atexit(), we want to continue cleanup. If it's being
+        // garbage-collected, the call is completely unpredictable from
+        // the consuming script's point of view. But what to do about this
+        // error?? For now, just log it.
+        LL_WARNS("Lua") << "setdtor(" << std::quoted(desc) << ") error: "
```

Comment:
Maybe, we should show script name here too, if the lua error is not thrown.

Reply:
Really good point! I've added it to other Lua log messages too.